### ISSUE TITLE
Bind JSON and URI 

### ISSUE THREAD

**Issue by montanaflynn** [gin-gonic/gin#1824]
## Description

I'm trying to bind JSON and a URI in one request but havn't found out how to yet. 

Here's my struct:

```go
type ResetPassword struct {
	Code     string `uri:"code" binding:"required"`
	Password string `json:"password" binding:"required,min=6,max=20"`
}
```

I'm using `c.Bind()`:

```go
func verifyPasswordReset(c *gin.Context) {
	var data ResetPassword
	if err := c.Bind(&data); err != nil {
		c.JSON(400, gin.H{"error": err})
		return
	}
	fmt.Printf("%+v\n", data)
	c.JSON(http.StatusOK, data)
}
```

When I make this request it says the code is required:

```
curl -H "content-type:application/json" -d '{"password":"hunter15"}' "localhost:8080/user/password_reset/jksakhdkh"                                                                     Sun Mar 24 14:29:27 2019
{"error":{".Code":{"FieldNamespace":".Code","NameNamespace":"Code","Field":"Code","Name":"Code","Tag":"required","ActualTag":"required","Kind":24,"Type":{},"Param":"","Value":""}}}
```

I found a related issue #811 but it wasn't actually solved before being closed.

**Comment by thinkerou** [comment]
you should split bind uri and json to two struct.

**Comment by montanaflynn** [comment]
@thinkerou ok, I split it into two structs and used the respective binds:

```go
	if err := c.ShouldBindUri(&dataURI); err != nil {
		c.JSON(400, gin.H{"error": err})
		return
	}

	if err := c.ShouldBindJSON(&dataJSON); err != nil {
		c.JSON(400, gin.H{"error": err})
		return
	}
```

That works fine and removes my workaround of using `c.Params.ByName("code")`, thanks!

**Comment by m430** [comment]
@thinkerou Why not use the uri and form tag together? I think it would be convenient.

**Comment by invisibleDesigner** [comment]
no, I want to use it in one  struct

**Comment by chengr4** [comment]
At 2023, we still should split it into two structs?

**Comment by TaylorDurden** [comment]
> At 2023, we still should split it into two structs?
Here is a example:

type updateAccountRequest struct {
	ID      int64 `uri:"id" binding:"required,min=1"`
	Balance int64 `json:"balance"`
}

func (server *Server) updateAccountHandler(ctx *gin.Context) {
	var req updateAccountRequest
	if err := ctx.ShouldBindUri(&req); err != nil {
		ctx.JSON(http.StatusBadRequest, errorResponse(err))
		return
	}

	if err := ctx.ShouldBindJSON(&req); err != nil {
		ctx.JSON(http.StatusBadRequest, errorResponse(err))
		return
	}
	arg := db.UpdateAccountParams{
		ID:      req.ID,
		Balance: req.Balance,
	}
	updatedAccount, err := server.store.UpdateAccount(ctx, arg)
	if err != nil {
		if err == sql.ErrNoRows {
			ctx.JSON(http.StatusNotFound, errorResponse(err))
			return
		}
		ctx.JSON(http.StatusInternalServerError, errorResponse(err))
		return
	}

	ctx.JSON(http.StatusOK, updatedAccount)
}

**Comment by QingShan-Xu** [comment]
> Balance int64 `json:"balance"`

If you set Balance int64 `json:"balance" binding:"required"`, it will still fail because shoudBindUri binds json:"balance" but there is no balance in the uri.

**Comment by xelarion** [comment]
```go
// Example usage:

func verifyPasswordReset(c *gin.Context) {
	var data ResetPassword
	if err := binding.Bind(c, &req, binding.URI, binding.JSON); err != nil {
		c.JSON(400, gin.H{"error": err})
		return
	}
	
	fmt.Printf("%+v\n", data)
	c.JSON(http.StatusOK, data)
}
```

package sources:

```go
// Package binding provides enhanced binding utilities that extend Gin's binding capabilities.
// It solves the common problem of binding data from multiple sources (URI, JSON, etc.)
// without validation errors caused by partially filled structs.
package binding

import (
	"sync"

	"github.com/gin-gonic/gin"
	ginbinding "github.com/gin-gonic/gin/binding"
)

// Internal package variables
var (
	// originalValidator stores the original Gin validator for use in final validation
	originalValidator ginbinding.StructValidator

	// initOnce ensures the validator is captured only once
	initOnce sync.Once
)

// init captures the original validator and sets the global validator to nil.
// This happens at package import time and only once.
func init() {
	// Use Once to ensure this only happens once even if multiple goroutines enter simultaneously
	initOnce.Do(func() {
		// Save original validator
		originalValidator = ginbinding.Validator

		// Set global validator to nil to skip validation during binding
		ginbinding.Validator = nil
	})
}

// Bind processes data from multiple sources without validating between each step,
// then performs a single validation at the end.
//
// This solves the problem where binding from one source (e.g., URI) fails validation
// because fields from another source (e.g., JSON) haven't been bound yet.
//
// Example usage:
//
//	var req UserUpdateRequest
//	if err := binding.Bind(c, &req, binding.URI, binding.JSON); err != nil {
//	    c.JSON(http.StatusBadRequest, ErrorResponse{Message: err.Error()})
//	    return
//	}
func Bind(c *gin.Context, obj interface{}, bindFuncs ...func(*gin.Context, interface{}) error) error {
	// Execute all binding functions (validation is already disabled)
	for _, bindFunc := range bindFuncs {
		if err := bindFunc(c, obj); err != nil {
			return err // Returns parsing errors but not validation errors
		}
	}

	// Perform validation manually after all bindings are complete
	if originalValidator != nil {
		return originalValidator.ValidateStruct(obj)
	}

	return nil
}

// Common binding functions for use with Bind

// URI binds URI parameters to the given object.
// Uses Gin's ShouldBindUri but without validation.
func URI(c *gin.Context, obj interface{}) error {
	return c.ShouldBindUri(obj)
}

// JSON binds JSON body to the given object.
// Uses Gin's ShouldBindJSON but without validation.
func JSON(c *gin.Context, obj interface{}) error {
	return c.ShouldBindJSON(obj)
}

```

**Comment by rogersm** [comment]
Thanks for this solution. It allows me to have a single Bind call independently on how data is submitted by client.
